In [53]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import scipy.stats

### 2. Given the following confusion matrix, evaluate (by hand) the model's performance.

|               | pred dog   | pred cat   |
|:------------  |-----------:|-----------:|
| actual dog    |         46 |         7  |
| actual cat    |         13 |         34 |


- In the context of this problem, what is a false positive?
- In the context of this problem, what is a false negative?
- How would you describe this model?

##### - Positive case: Dogs
##### - Negative case: Cats

##### - False positive: predicting dogs and actuals are cats (13)
##### - False negative: predicting cats and actuals are dogs (7)

In [27]:
# - Description of the model: this model is a DOG classifier 
# because DOGS are the positive prediction

tp = true_positive = 46
tn = true_negative = 34
fp = false_positive = 13
fn = false_negative = 7

accuracy = (tp + tn) / (tp + tn + fp + fn)
recall = tp / (tp + fn)
precision = tp / (tp + fp)

print(f'Accuracy: {accuracy}\n')
print(f'Recall: {round(recall, 2)}\n')
print(f'Precision: {round(precision, 2)}')

Accuracy: 0.8

Recall: 0.87

Precision: 0.78


### 3. You are working as a datascientist for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.

Use the predictions dataset and pandas to help answer the following questions:

- An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?
- Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

##      a.

In [12]:
# Getting to know the content of the DataFrame

c3 = pd.read_csv('c3.csv')
c3

actual     model1     model2     model3
0    No Defect  No Defect     Defect  No Defect
1    No Defect  No Defect     Defect     Defect
2    No Defect  No Defect     Defect  No Defect
3    No Defect     Defect     Defect     Defect
4    No Defect  No Defect     Defect  No Defect
..         ...        ...        ...        ...
195  No Defect  No Defect     Defect     Defect
196     Defect     Defect  No Defect  No Defect
197  No Defect  No Defect  No Defect  No Defect
198  No Defect  No Defect     Defect     Defect
199  No Defect  No Defect  No Defect     Defect

[200 rows x 4 columns]

In [31]:
# Looking for null values and their count

c3.isna().value_counts()

actual  model1  model2  model3
False   False   False   False     200
dtype: int64

In [28]:
# Discovering the count of each variable in actual data

c3.actual.value_counts()

No Defect    184
Defect        16
Name: actual, dtype: int64

**Summary for myself:**
- Mode or baseline = 184 (no defects, the number that is repeated the most)
- type of model = a defect classifier (we are looking for defects)
- metric = Recall (recall measure the misses and I don't want to miss the defects)

##### - Since the goal is to identify as many defects as possible, defect is the classifier 
##### and recall the metric because it checks for how many defects are actually flagged

In [41]:
# Doing a crosstab to get TP, TN, FP, and FN for model1

pd.crosstab(c3.actual, c3.model1)

model1     Defect  No Defect
actual                      
Defect          8          8
No Defect       2        182

In [42]:
# Doing a crosstab to get TP, TN, FP, and FN for model2

pd.crosstab(c3.actual, c3.model2)

model2     Defect  No Defect
actual                      
Defect          9          7
No Defect      81        103

In [43]:
# Doing a crosstab to get TP, TN, FP, and FN for model3

pd.crosstab(c3.actual, c3.model3)

model3     Defect  No Defect
actual                      
Defect         13          3
No Defect      86         98

In [48]:
# Calculating reacll for all three

recall_1 = 8/(8+7)

recall_2 = 9/(9+3)

recall_3 = 13/(13+3)


print(f'Recall_1: {round(recall_1, 2)}\n')
print(f'Recall_1: {recall_2}\n')
print(f'Recall_1: {round(recall_3, 2)}\n')

Recall_1: 0.53

Recall_1: 0.75

Recall_1: 0.81



##### The recall rate is higher for model3. I suggest the internal team choose it

### b.

In [70]:
# Because I am looking to accurately predict TP
# and avoid FP at all cost, FP have a high price 
# here so I need to use precision

precision_1 = 8/(8+2)

precision_2 = 9/(9+81)

precision_3 = 13/(13+86)


print(f'Precision_1: {precision_1}\n')
print(f'Precision_2: {precision_2}\n')
print(f'Precision_3: {round(precision_3, 2)}\n')

Precision_1: 0.8

Precision_2: 0.1

Precision_3: 0.13



##### The precision is higher for model1. This is the best model to avoid giving our vacation to the wrong clients

### 4. You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

Several models have already been developed with the data, and you can find their results here.

Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:

In [109]:
# Getting the data from the csv file that was downloaded to my computer when I clicked on it

data2 = pd.read_csv('gives_you_paws.csv')
data2.head()

actual model1 model2 model3 model4
0    cat    cat    dog    cat    dog
1    dog    dog    cat    cat    dog
2    dog    cat    cat    cat    dog
3    dog    dog    dog    cat    dog
4    cat    cat    cat    dog    dog

In [64]:
# Getting the count of the values.
# Click on the three dots to open the results box

data2.value_counts()

actual  model1  model2  model3  model4
dog     dog     cat     cat     dog       644
                        dog     dog       632
                dog     dog     dog       625
                        cat     dog       590
cat     cat     cat     cat     dog       412
                        dog     dog       409
                        cat     cat       243
                        dog     cat       198
dog     cat     dog     dog     dog       170
                cat     dog     dog       155
                        cat     dog       150
                dog     cat     dog       144
cat     dog     cat     dog     dog       100
                        cat     dog        96
        cat     dog     dog     dog        57
        dog     cat     cat     cat        50
                        dog     cat        47
        cat     dog     cat     dog        47
dog     dog     cat     dog     cat        40
                dog     cat     cat        30
cat     cat     dog     cat     cat      

#### a. In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?

In [110]:
# Getting a statistical summary of the data

data2.describe().T

count unique  top  freq
actual  5000      2  dog  3254
model1  5000      2  dog  2937
model2  5000      2  cat  3212
model3  5000      2  dog  2508
model4  5000      2  dog  4253

In [111]:
# Getting a specific count of the the actual column for baseline

data2.actual.value_counts()

dog    3254
cat    1746
Name: actual, dtype: int64

In [120]:
# Creating a baseline column

data2["baseline_prediction"] = 'dog'
data2.head()

actual model1 model2 model3 model4 baseline baseline_prediction
0    cat    cat    dog    cat    dog      dog                 dog
1    dog    dog    cat    cat    dog      dog                 dog
2    dog    cat    cat    cat    dog      dog                 dog
3    dog    dog    dog    cat    dog      dog                 dog
4    cat    cat    cat    dog    dog      dog                 dog

In [140]:
# Removing the result of previous column manipulation

data2 = data2.drop(['baseline'], axis = 1)
data2.head()

actual model1 model2 model3 model4 baseline_prediction
0    cat    cat    dog    cat    dog                 dog
1    dog    dog    cat    cat    dog                 dog
2    dog    cat    cat    cat    dog                 dog
3    dog    dog    dog    cat    dog                 dog
4    cat    cat    cat    dog    dog                 dog

In [157]:
# Groupng the models in one variable and selecting all but the actuals

models = list(data2.columns)
models = models[1:]
models

['model1', 'model2', 'model3', 'model4', 'baseline_prediction']

In [158]:
# Creating an output for my model through a dictionary

output = {}
for model in models:
    accuracy = (data2.actual == data2[model]).mean()
    output.update({model:accuracy})
    
output

{'model1': 0.8074,
 'model2': 0.6304,
 'model3': 0.5096,
 'model4': 0.7426,
 'baseline_prediction': 0.6508}

In [159]:
# Practicing for myself by trying to the recall for all the models
# As expected recall of the baseline is 100% but model 4 is doing very good followed by model 1 as well

subset = data2[data2.actual == 'dog']

output2 = {}
for model in models:
    recall = (subset[model] == subset.actual).mean()
    output2.update({model:recall})
    
output2

{'model1': 0.803318992009834,
 'model2': 0.49078057775046097,
 'model3': 0.5086047940995697,
 'model4': 0.9557467732022127,
 'baseline_prediction': 1.0}

In [131]:
accuracy_results = pd.DataFrame(output.items(), columns = ['model', 'accuracy'] )
accuracy_results

model  accuracy
0    model1  0.803319
1    model2  0.490781
2    model3  0.508605
3    model4  0.955747
4  baseline  1.000000

In [132]:
recall_results = pd.DataFrame(output2.items(), columns = ['model', 'recall'] )
recall_results

model    recall
0    model1  0.803319
1    model2  0.490781
2    model3  0.508605
3    model4  0.955747
4  baseline  1.000000

In [134]:
merged_models = pd.merge(accuracy_results, recall_results)
merged_models

model  accuracy    recall
0    model1  0.803319  0.803319
1    model2  0.490781  0.490781
2    model3  0.508605  0.508605
3    model4  0.955747  0.955747
4  baseline  1.000000  1.000000

##### Model1 and 4 have a higher accuracy than the baseline model

### b. Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recomend for Phase I? For Phase II?

### Phase 1

Considering the dog team

Dog is the positive

Cat is the negative

Indicator is Recall

In [143]:
subset_dog = data2[data2.actual == 'dog']
subset_dog.head()

actual model1 model2 model3 model4 baseline_prediction
1    dog    dog    cat    cat    dog                 dog
2    dog    cat    cat    cat    dog                 dog
3    dog    dog    dog    cat    dog                 dog
5    dog    dog    dog    dog    dog                 dog
8    dog    dog    cat    dog    dog                 dog

In [154]:
models_dog = list(subset_dog.columns)
models_dog = models_dog[1:]
models_dog

['model1', 'model2', 'model3', 'model4', 'baseline_prediction']

In [161]:
output_dog = {}
for model in models_dog:
    recall = (subset_dog.actual == subset_dog[model]).mean()
    output_dog.update({model:recall})
    
output_dog

{'model1': 0.803318992009834,
 'model2': 0.49078057775046097,
 'model3': 0.5086047940995697,
 'model4': 0.9557467732022127,
 'baseline_prediction': 1.0}

**The best model is model4 with 96% Recall score**

### Phase 2

Indicator is Precision

In [163]:
data2.head()

actual model1 model2 model3 model4 baseline_prediction
0    cat    cat    dog    cat    dog                 dog
1    dog    dog    cat    cat    dog                 dog
2    dog    cat    cat    cat    dog                 dog
3    dog    dog    dog    cat    dog                 dog
4    cat    cat    cat    dog    dog                 dog

In [165]:
# Creating a subset for each model in order to get precision

subset1 = data2[data2.model1 == 'dog']
subset2 = data2[data2.model2 == 'dog']
subset3 = data2[data2.model3 == 'dog']
subset4 = data2[data2.model4 == 'dog']

In [166]:
# Model1 precision

(subset1.actual == subset1.model1).mean()

0.8900238338440586

In [167]:
# Model2 precision

(subset2.actual == subset2.model2).mean()

0.8931767337807607

In [169]:
# Model3 precision

(subset3.actual == subset3.model3).mean()

0.6598883572567783

In [170]:
# Model4 precision

(subset4.actual == subset4.model4).mean()

0.7312485304490948

**Best precision is Model2 with 89.31%**

### c. Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recomend for Phase I? For Phase II?

Cat is positive

⚠️ Review before Tuesday mormning ⚠️

### Phase 1

In [172]:
data2.head()

actual model1 model2 model3 model4 baseline_prediction
0    cat    cat    dog    cat    dog                 dog
1    dog    dog    cat    cat    dog                 dog
2    dog    cat    cat    cat    dog                 dog
3    dog    dog    dog    cat    dog                 dog
4    cat    cat    cat    dog    dog                 dog

In [175]:
data2 = data2.drop(['baseline_prediction'], axis = 1)
data2

actual model1 model2 model3 model4
0       cat    cat    dog    cat    dog
1       dog    dog    cat    cat    dog
2       dog    cat    cat    cat    dog
3       dog    dog    dog    cat    dog
4       cat    cat    cat    dog    dog
...     ...    ...    ...    ...    ...
4995    dog    dog    dog    dog    dog
4996    dog    dog    cat    cat    dog
4997    dog    cat    cat    dog    dog
4998    cat    cat    cat    cat    dog
4999    dog    dog    dog    dog    dog

[5000 rows x 5 columns]

In [192]:
data2['baseline'] = 'dog'
data2.head()

actual model1 model2 model3 model4 baseline
0    cat    cat    dog    cat    dog      dog
1    dog    dog    cat    cat    dog      dog
2    dog    cat    cat    cat    dog      dog
3    dog    dog    dog    cat    dog      dog
4    cat    cat    cat    dog    dog      dog

In [193]:
subset_cat = data2[data2.actual == 'cat']
subset_cat.head()

actual model1 model2 model3 model4 baseline
0     cat    cat    dog    cat    dog      dog
4     cat    cat    cat    dog    dog      dog
6     cat    cat    cat    cat    dog      dog
7     cat    dog    cat    cat    dog      dog
11    cat    cat    dog    cat    cat      dog

In [194]:
models_cat = list(subset_cat.columns)
models_cat = models_cat[1:]
models_cat

['model1', 'model2', 'model3', 'model4', 'baseline']

In [196]:
output_cat = {}
for model in models_cat:
    recall = (subset_cat.actual == subset_cat[model]).mean()
    output_cat.update({model:recall})
    
output_cat

{'model1': 0.8150057273768614,
 'model2': 0.8906071019473081,
 'model3': 0.5114547537227949,
 'model4': 0.34536082474226804,
 'baseline': 0.0}

**Best model is Model2 89.06%**

### Phase 2

Indicator is precision

In [197]:
subset_cat1 = data2[data2.model1 == 'dog']
subset_cat2 = data2[data2.model2 == 'dog']
subset_cat3 = data2[data2.model3 == 'dog']
subset_cat4 = data2[data2.model4 == 'dog']

In [198]:
# Model1 precision

(subset1.actual == subset1.model1).mean()

0.8900238338440586

In [199]:
# Model2 precision

(subset2.actual == subset2.model2).mean()

0.8931767337807607

In [200]:
# Model1 precision

(subset3.actual == subset3.model3).mean()

0.6598883572567783

In [186]:
# Model1 precision

(subset4.actual == subset4.model4).mean()

0.7312485304490948

### 5. Follow the links below to read the documentation about each function, then apply those functions to the data from the previous problem.